[joy-ui][Drawer] Add Drawer component

mnajdova opened the pull request:
closes #36292

Adding Drawer related components in Joy UI. I started with adding a `Drawer` component - which is the equivalent of the temporary drawer in Material UI. I decided not to add the `permanent` variant version, because in my opinion, these two versions of the component are too different to justify them being bundled in one component - if developers need a permanent drawer, why bundle all the modal-related code with it?

After this PR:

- Add PermanentDrawer (Sidebar) component - we should have an integration demo about how the two can be used together.
- Add SwipeableDrawer component
- Add PersistentDrawer component

mnajdova commented on packages/mui-joy/src/Modal/Modal.tsx:
```diff
@@ -117,6 +117,8 @@ const Modal = React.forwardRef(function ModalU(inProps, ref) {
     slots = {},
     slotProps = {},
+    // @ts-ignore TODO: Add it in the types - how can we make the component extendable?
+    ownerState: ownerStateProp,
```
This is one things that we will need to resolve if we decide to go with component composition approach.

mnajdova commented on packages/mui-joy/src/Drawer/drawerClasses.ts:
```diff
@@ -0,0 +1,18 @@
+  /** Class name applied to the root element. */
+  root: string;
+  /** Class name applied to the sheets element. */
+  sheet: string;
```
Not happy about this name.

mnajdova commented on packages/mui-joy/src/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,213 @@
+      component={component}
+      slots={modalSlots}
+      // @ts-ignore ownerState is propagated to the Modal component
+      slotProps={modalSlotProps}
```
Another thing to be resolved if using compoentn composition.

mnajdova commented on packages/mui-joy/src/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,213 @@
+ * - [Drawer API](https://mui.com/joy-ui/api/drawer/)
+ * - inherits [Modal API](https://mui.com/joy-ui/api/modal/)
+ */
+const Drawer = React.forwardRef(function Drawer(inProps, ref) {
```
It's less than 60 lines of code with component composition, in comparison with the modal for e.g. that has around 200 lines of code. This is only to compare the complexity

mnajdova commented on packages/mui-joy/src/Drawer/Drawer.test.tsx:
```diff
@@ -0,0 +1,58 @@
+import Modal, { modalClasses } from '@mui/joy/Modal';
+import Drawer, { drawerClasses as classes } from '@mui/joy/Drawer';
+
+describe('<Drawer />', () => {
```
We'll need to add more tests.

siriwatknp commented:
The reason I don't want to build a component on top of another component is to avoid the issues we have seen in Material UI. This is what I meant by build a component on top of another component:

```js
// Modal.tsx
export default
```

mnajdova commented:
> For the case of Drawer, we want to reuse the modal logic so why not create a hook useModal (we could just add it in Joy UI internally first and let Base UI take over later). Then reuse the useModal in both Modal and Drawer.
> 
> For the styles, export the styled component from the Modal and reuses in Drawer (this approach is already been using in other Joy UI components like Autocomplete).

I can try as a first step to isolate the `useModal` hook. My worry is that much of the Modal logic is happening in how elements are being rendered:

```
    <CloseModalContext.Provider value={onClose}>
      <Portal ref={handlePortalRef} container={container} disablePortal={disablePortal}>
        {/*
         * Marking an element with the role presentation indicates to assistive technology
         * that this element should be ignored; it exists to support the web application and
         * is not meant for humans to interact with directly.
         * https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-static-element-interactions.md
         */}
        <SlotRoot {...rootProps}>
          {!hideBackdrop ? <SlotBackdrop {...backdropProps} /> : null}
          <FocusTrap
            disableEnforceFocus={disableEnforceFocus}
            disableAutoFocus={disableAutoFocus}
            disableRestoreFocus={disableRestoreFocus}
            isEnabled={isTopModal}
            open={open}
          >
            {React.Children.only(children) &&
              React.cloneElement(children, {
                ...(children.props.tabIndex === undefined && {
                  tabIndex: -1,
                }),
              })}
          </FocusTrap>
        </SlotRoot>
      </Portal>
    </CloseModalContext.Provider>
```

If we ever need to do a fix on this logic, we will need to apply it in multiple places, while if we re-use the component it would live only in one place. Let me create the `useModal` hook first and see what can be abstracted.

mnajdova commented on packages/mui-joy/src/Modal/modalClasses.ts:
```diff
@@ -3,6 +3,8 @@ import { generateUtilityClass, generateUtilityClasses } from '../className';
   /** Class name applied to the root element. */
   root: string;
+  /** Class name applied to the root element when open is false. */
+  hidden: string;
```
Missing class name

siriwatknp commented on packages/mui-joy/src/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,306 @@
+                isEnabled={isTopModal}
+                open={open}
+              >
+                <SlotContent {...contentProps} tabIndex={contentProps.tabIndex ?? -1}>
```
This line is strange to me. Can it be included in the `additionalProps`?

```js
const [SlotContent, contentProps] = useSlot({
  …,
  additionalProps: {
    tabIndex: tabIndex ?? -1
  }
})
```

mnajdova replied:
Yep, refactored and added test for it.

siriwatknp commented on packages/mui-joy/src/DrawerContent/DrawerContent.tsx:
```diff
@@ -0,0 +1,228 @@
+'use client';
```
Maybe the DrawerContent is not needed as well.

mnajdova replied:
Agreed, removed.

siriwatknp commented:
@mnajdova For the structure, I think `DrawerContent` can be replaced with Drawer's content slot to simplify the implementation (remove unnecessary React context). I'd expect to use the Drawer like this:

```js
<Drawer>
  <Typography>Title</Typography>
  …
</Drawer>
```

to customize the content use `slotProps.content`.

mnajdova commented:
> @mnajdova For the structure, I think DrawerContent can be replaced with the Drawer's content slot to simplify the implementation (remove unnecessary React context). I'd expect to use the Drawer like this:

I was considering this as well, I mainly created this component following the `ModalDialog` paradigm we have, but yeah, it makes sense to have it as slot too.

> Should PermanentDrawer be excluded from this PR? It would reduce the size of the PR.

Yeah, let me extract these changes in a new branch. Btw, considering we will have more `*Drawer` components, should I rename the default one to `TemporaryDrawer` or keep `Drawer` for simplicity, considering this is (at least in my opinion), the componentI first have in mind when I think about a Drawer component. Maybe we can loop @samuelsycamore too on the name.

siriwatknp commented:
> I decided to keep the `PermanentDrawer` as part of the initial effort - the docs would be basically empty without it - I feel this is basic requirement for the Drawer component - I just decided to extract it to a new component - for e.g. in Material UI it is defined by a prop on the `Drawer`.

Got your point, we can include another opinion from @samuelsycamore or @zanivan about the component. To be honest, I don't think we should follow Material UI on this, I feel that a `TemporaryDrawer` is the definition of a Drawer, and `PermanentDrawer` should not exist because using Flexbox or CSS Grid is a lot better to deal with the layout.

Some benchmark:
- https://chakra-ui.com/docs/components/drawer/usage
- https://ant.design/components/drawer
- https://ui.shadcn.com/docs/components/sheet

zanivan commented:
Let me chime in: I think the default should be `Drawer`, and the default behavior should be similar to `TemporaryDrawer` because, as @siriwatknp described, this is the definition of a drawer, and it's what's more commonly seen. `SwipableDrawer` could be a good prop to be added, but I believe it's not exactly a must-have feature, and we could work around it with a way of doing this in a demo, and not properly a built-in prop. 

However, I'd love to have something similar to the `PermanentDrawer`, mostly to be used as a `sideBar` or `sideMenu`—it indeed can be achieved in many ways, but it'd be great to have a similar component that provides some behaviors like collapsing, closing, etc. 

That said, I believe it's out of the scope of a `Drawer` itself, since the main usage is to be a side navigation, that doesn't fit exactly with the concept of a drawer.

mnajdova commented:
I scoped the PR to only add the Drawer component, I can open an issue for the permanent drawer variant (we could likely call it a sidebar, it's much more used than permanent drawer), but we can discuss this on the issue. I have the implementation already, so it would be moslty about how we want to name it and document it.

siriwatknp commented:
> However, I'd love to have something similar to the `PermanentDrawer`, mostly to be used as a `sideBar` or `sideMenu`—it indeed can be achieved in many ways, but it'd be great to have a similar component that provides some behaviors like collapsing, closing, etc.

I have the same thought. The permanent drawer is a sidebar most of the time, which needs to be responsive, and integrated with other parts e.g Header, Main, Footer, and as @zanivan pointed out, collapsable at some point.

We should plan ahead for the Sidebar. I believe that the Sidebar should be in the system package which uses the zero-runtime engine that @brijeshb42 is working on. This could be a great opportunity for us to create layout components that are 100% server components.

siriwatknp commented on docs/data/joy/components/drawer/drawer.md:
```diff
@@ -1,26 +1,23 @@
+The Drawer can be cancelled by clicking the overlay or pressing the Esc key.
+It closes when an item is selected, handled by controlling the `open` prop.
+
+{{"demo": "DrawerBasic.js"}}
```
```suggestion

## Customization

### Anchor position

{{"demo": "DrawerAnchor.js"}}
```

I think the DrawerBasic can be simpler with the default anchor. Maybe an 
`<IconButton><MenuIcon /></IconButton>` that toggle the Drawer.

and then add another demo in the customization for changing the anchor.

mnajdova replied:
Restructured a bit the demos:
- the basic containes simple app bar & sandwich that open drawer - this is the most common use-case
- the customization has demos for the different anchor & sizes that can be used on the drawer.

siriwatknp commented on docs/data/joy/components/drawer/DrawerUsage.js:
```diff
@@ -0,0 +1,125 @@
+        {['Inbox', 'Starred', 'Send email', 'Drafts'].map((text, index) => (
+          <ListItem key={text}>
+            <ListItemButton>
+              <ListItemButton>
```
```suggestion
```

duplicate ListItemButton

siriwatknp commented on packages/mui-joy/src/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,399 @@
+  slot: 'Content',
+  overridesResolver: (props, styles) => styles.root,
+})<{ ownerState: DrawerOwnerState }>(({ theme, ownerState }) => ({
+  '--DrawerContent-radius': 0,
```
```suggestion
```
I think this is not required.

siriwatknp commented on packages/mui-joy/src/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,399 @@
+  }),
+  boxShadow: theme.shadow.md,
+  borderRadius: 'var(--DrawerContent-radius)',
+  fontFamily: theme.vars.fontFamily.body,
+  lineHeight: theme.vars.lineHeight.md,
```
Should it be:

```suggestion
  …theme.typography[`body-${ownerState.size}`],
```

other components are using this pattern.

siriwatknp commented on packages/mui-joy/src/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,399 @@
+  }),
+  height: ownerState.anchor!.match(/(left|right)/) ? '100%' : 'auto',
+  width: ownerState.anchor!.match(/(top|bottom)/) ? '100vw' : 'auto',
+  transition: 'transform 0.3s ease',
```
tag @zanivan for reviewing the transition/animation.

marcpachecog replied:
Just for clarification, you decided to remove the transitions of all components in favor of adding them in a "plugin" fashion way, but the Accordion and now the Drawer also have them.

zanivan replied:
Just sharing my thoughts: Though I like the idea of having a more robust feature for animations and transitions, I think we can make some smaller additions meanwhile. This is mostly for the scenarios where a) it's very common to have animations, and b) for the newly introduced components. 
As for the existing components, we could potentially handle them on a case-by-case basis if they fall into category a), like `switch` for instance. However, if it gets overly complex, it might be better to hold off until the “animation plugin” is available.

mnajdova replied:
Update the transition looks better.

> Just for clarification, you decided to remove the transitions of all components in favor of adding them in a "plugin" fashion way, but the Accordion and now the Drawer also have them.

This is a bigger discussion, we should probably move this to the issue about it - https://github.com/mui/material-ui/issues/36695. In general in my opinion we should have a minimal transitions by default, that will make the components feel great - in addition we can have mechanisms for extending these defaults.

siriwatknp replied:
@marcpachecog We will add a transition demo to give an example on how to customize it.

siriwatknp commented on packages/mui-joy/src/Drawer/Drawer.tsx:
```diff
@@ -0,0 +1,399 @@
+  return ['left', 'right'].indexOf(anchor!) !== -1;
+}
+
+export function getAnchor(theme: Theme, anchor: DrawerProps['anchor']) {
+  // @ts-ignore TODO: How is the direction set in Joy UI?
+  return theme.direction === 'rtl' && isHorizontal(anchor) ? oppositeDirection[anchor] : anchor;
+}
```
Joy does not support RTL at the moment, we could remove it for now?

siriwatknp commented:
@mnajdova

In https://github.com/mui/material-ui/pull/38169/commits/5c1eebc9cc7e3802468337c3c99d5b058cad2902, I did the same for the Root and Content slot by exporting the styled-components from the Modal. The reason to use the styled-components, not the slot component is to prevent `theme.styleOverrides` inheritance.

In https://github.com/mui/material-ui/pull/38169/commits/8768e6cf3392a634f933073822213fc3f6cbc889, I removed the variables that are 1-to-1 to the CSS properties and define the width, height variables for each size (we can discuss if those variables make sense or not)

In https://github.com/mui/material-ui/pull/38169/commits/bd816084dad210925ada9944ca453d901daf7fbc, I think the drawer should support multiple children. Furthermore, I think the `getTransitionProps` should not be in `useModal`. The API is too specific to Transition component, may be it should be separate into another hook.

The rest are demos and examples.

mnajdova commented:
> In https://github.com/mui/material-ui/commit/5c1eebc9cc7e3802468337c3c99d5b058cad2902, I did the same for the Root and Content slot by exporting the styled-components from the Modal. The reason to use the styled-components, not the slot component is to prevent theme.styleOverrides inheritance.

Nice, this pattern makes sense 👍 

> In https://github.com/mui/material-ui/commit/8768e6cf3392a634f933073822213fc3f6cbc889, I removed the variables that are 1-to-1 to the CSS properties and define the width, height variables for each size (we can discuss if those variables make sense or not)

Yep, I was thinking whether we should go with this route - the code I had was mostly copied from the modal dialog.

> In https://github.com/mui/material-ui/commit/bd816084dad210925ada9944ca453d901daf7fbc, I think the drawer should support multiple children. Furthermore, I think the getTransitionProps should not be in useModal. The API is too specific to Transition component, may be it should be separate into another hook.

Makes total sense, we will likely change the way how transitions are handled in Base UI - this is kind of a legacy approach coming from Material UI.

zanivan commented:
There are some items that might be worth mentioning/discussing:
- The variants and colors are not being applied to the playground demo
- When switching modes, the iframe background sometimes appears and sometimes not 

| Light | Dark |
|--------|--------|
| ![Screenshot 2023-08-31 at 18 37 07](https://github.com/mui/material-ui/assets/37222944/c1d0398f-8a32-4b1f-b9a2-3fa015133199) | ![Screenshot 2023-08-31 at 18 38 19](https://github.com/mui/material-ui/assets/37222944/5a1e77e6-14e5-4cc6-92b0-98ff4bfa02ce) | 
- It's impossible to dismiss the `lg` modal on mobile
<img width="200" alt="Screenshot" src="https://github.com/mui/material-ui/assets/37222944/15bf4dda-113a-4206-b05a-1488793d3496">

---

I'm still getting up to speed on this, and you might already be aware of these things, so sorry about that. 
Just to let you know, I've started refining the demos designs. Should I wait a bit more, or keep working on them?

siriwatknp commented on docs/data/joy/components/drawer/drawer.md:
```diff
@@ -1,26 +1,55 @@
+The Drawer can be cancelled by clicking the overlay or pressing the Esc key.
+It closes when an item is selected, handled by controlling the `open` prop.
+
+{{"demo": "DrawerBasic.js", "iframe": true}}
```
Does it need to be iframe given that other demos are not? cc @zanivan

mnajdova replied:
I will change this demo to be similar to the others, I wanted to make it closer to showcase how it would look like in app, but it complicates things in the code itself - the usage of the document object etc.

mnajdova commented:
Sorry for the delay. Here are some answers:

>  The variants and colors are not being applied to the playground demo

Fixed, I must have broken this when I was removing the `DrawerContent` component. 

> When switching modes, the iframe background sometimes appears and sometimes not

I simplified this demo, no need for iframe, it complicates things when exporting to sandbox too.

> It's impossible to dismiss the lg modal on mobile

It's possible by clicking on some of the items. In practise, I wouldn't expect that someone would use large drawer on a mobile?

siriwatknp commented:
> It's impossible to dismiss the lg modal on mobile

> It's possible by clicking on some of the items. In practise, I wouldn't expect that someone would use large drawer on a mobile?

I added support for this by setting the width to `min(100vw - 2.5rem, drawer-width)`. With this, the drawer's width will never exceed the viewport's width which is commonly used.

siriwatknp reviewed (approved):
🚀🚀🚀

marcpachecog commented:
Thanks @siriwatknp , @mnajdova and 
@zanivan for your hard and meticulous work!

Usama-Tahir commented:
Great work. I have a question. When will it be available via npm?

Thanks

mnajdova commented:
> Great work. I have a question. When will it be available via npm?

Likely later today with the latest release :)
